### Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

In [1]:
import numpy as np
from scipy.stats import f

def f_test(array1, array2):
    var1 = np.var(array1, ddof=1)
    var2 = np.var(array2, ddof=1)
    n1 = len(array1)
    n2 = len(array2)
    F = var1 / var2
    p = f.cdf(F, n1 - 1, n2 - 1)
    return F, p

In [2]:
array1 = [1, 2, 3, 4, 5]
array2 = [6, 7, 8, 9, 10]
F, p = f_test(array1, array2)
print("F-value: ", F)
print("p-value: ", p)

F-value:  1.0
p-value:  0.5


### Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [3]:
from scipy.stats import f

def critical_f_value(df_num, df_den, alpha=0.05):
    
    crit_f = f.ppf(1 - alpha/2, df_num, df_den)
    return crit_f

In [4]:
critical_f_value(3, 16)

4.07682306196248

### Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F- value, degrees of freedom, and p-value for the test.

In [6]:
import numpy as np
from scipy.stats import f

# Set the random seed for reproducibility
np.random.seed(42)

# Generate random samples from two normal distributions with known variances
n1 = 50
n2 = 60
mean1 = 0
mean2 = 0
var1 = 4
var2 = 6
sample1 = np.random.normal(mean1, np.sqrt(var1), n1)
sample2 = np.random.normal(mean2, np.sqrt(var2), n2)

# Use an F-test to determine if the variances are equal
variance1 = np.var(sample1, ddof=1)
variance2 = np.var(sample2, ddof=1)
f_value = variance1/variance2 if variance1 > variance2 else variance2/variance1
df1 = n1-1
df2 = n2-1
p_value = f.sf(f_value, df1, df2) * 2

# Output the F-value, degrees of freedom, and p-value for the test
print("F-value: ", f_value)
print("Degrees of freedom: ", df1, ", ", df2)
print("p-value: ", p_value)

F-value:  1.2908650068588141
Degrees of freedom:  49 ,  59
p-value:  0.34691616680101056


### Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

In [8]:
import numpy as np
from scipy.stats import f

# Set the random seed for reproducibility
np.random.seed(42)

# Define the parameters
n1 = 12
n2 = 12
var1 = 10
var2 = 15
alpha = 0.05

# Generate random samples from two normal distributions with known variances
sample1 = np.random.normal(0, np.sqrt(var1), n1)
sample2 = np.random.normal(0, np.sqrt(var2), n2)

# Calculate the sample variances
s1_sq = np.var(sample1, ddof=1)
s2_sq = np.var(sample2, ddof=1)

# Calculate the F-statistic
f_stat = s1_sq / s2_sq

# Calculate the critical F-value
df1 = n1 - 1
df2 = n2 - 1
crit_f = f.ppf(1 - alpha / 2, df1, df2)

# Calculate the p-value
p_value = f.sf(f_stat, df1, df2) * 2

# Print the results
print("Sample variance of population 1: ", s1_sq)
print("Sample variance of population 2: ", s2_sq)
print("F-statistic: ", f_stat)
print("Critical F-value: ", crit_f)
print("p-value: ", p_value)

if f_stat > crit_f:
    print("Reject the null hypothesis.")
else:
    print("Fail to reject the null hypothesis.")

Sample variance of population 1:  5.538947904274748
Sample variance of population 2:  14.990851102225099
F-statistic:  0.3694885544859157
Critical F-value:  3.473699051085809
p-value:  1.8866181475724586
Fail to reject the null hypothesis.


### Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

In [9]:
from scipy.stats import f

# Define the parameters
n = 25
sigma_sq = 0.005
s_sq = 0.006
alpha = 0.01

# Calculate the F-statistic
f_stat = s_sq / sigma_sq

# Calculate the critical F-value
df1 = n - 1
df2 = 1
crit_f = f.ppf(1 - alpha / 2, df1, df2)

# Calculate the p-value
p_value = f.sf(f_stat, df1, df2) * 2

# Print the results
print("Sample variance: ", s_sq)
print("Population variance: ", sigma_sq)
print("F-statistic: ", f_stat)
print("Critical F-value: ", crit_f)
print("p-value: ", p_value)

if f_stat > crit_f:
    print("Reject the null hypothesis.")
else:
    print("Fail to reject the null hypothesis.")

Sample variance:  0.006
Population variance:  0.005
F-statistic:  1.2
Critical F-value:  24939.565259943236
p-value:  1.259219923991872
Fail to reject the null hypothesis.


### Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

In [10]:
def f_dist_mean_var(df1, df2):
    
    mean = df2 / (df2 - 2)
    variance = (2 * df2**2 * (df1 + df2 - 2)) / (df1 * (df2 - 2)**2 * (df2 - 4))
    return mean, variance

In [11]:
mean, variance = f_dist_mean_var(5, 10)
print("Mean: ", mean)
print("Variance: ", variance)

Mean:  1.25
Variance:  1.3541666666666667


### Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

In [12]:
import scipy.stats as stats

# Sample sizes and variances
n1 = 10
s1_squared = 25
n2 = 15
s2_squared = 20

# Calculate the F-statistic
F = s1_squared / s2_squared

# Calculate the degrees of freedom
df1 = n1 - 1
df2 = n2 - 1

# Find the critical value of the F-statistic
alpha = 0.10
f_critical = stats.f.ppf(alpha/2, df1, df2)

# Calculate the p-value
p_value = 2 * (1 - stats.f.cdf(F, df1, df2))

# Compare F to the critical value
if F < f_critical or F > 1/f_critical:
    print("Reject the null hypothesis")
else:
    print("Fail to reject the null hypothesis")

# Output the F-statistic, critical value, and p-value
print("F-statistic: ", F)
print("Critical value: ", f_critical)
print("p-value: ", p_value)

Fail to reject the null hypothesis
F-statistic:  1.25
Critical value:  0.3305268601412525
p-value:  0.6832194382585952


### Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

#### Sample variance of Restaurant A:

In [13]:
x = [24, 25, 28, 23, 22, 20, 27]
n = len(x)
mean = sum(x) / n
variance = sum((xi - mean) ** 2 for xi in x) / (n - 1)
print(variance)

7.80952380952381


#### Sample variance of Restaurant B:

In [14]:
x = [31, 33, 35, 30, 32, 36]
n = len(x)
mean = sum(x) / n
variance = sum((xi - mean) ** 2 for xi in x) / (n - 1)
print(variance)

5.366666666666667


In [15]:
from scipy.stats import f

s1_sq = 7.80952380952381
s2_sq = 5.366666666666667
df1 = len(x) - 1
df2 = len(x) - 1

F = s1_sq / s2_sq
p_value = f.cdf(F, df1, df2)
print(p_value)

0.65468119327068


### Since the p-value (0.654) is greater than the significance level of 0.05, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the waiting times at the two restaurants are significantly different at the 5% significance level.

### Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

###### Sample variance of Group A:

In [16]:
x = [80, 85, 90, 92, 87, 83]
n = len(x)
mean = sum(x) / n
variance = sum((xi - mean) ** 2 for xi in x) / (n - 1)
print(variance)

19.76666666666667


###### Sample variance of Group B:

In [17]:
x = [75, 78, 82, 79, 81, 84]
n = len(x)
mean = sum(x) / n
variance = sum((xi - mean) ** 2 for xi in x) / (n - 1)
print(variance)

10.166666666666666


In [19]:
from scipy.stats import f

s1_sq = 19.76666666666667
s2_sq = 10.166666666666666
df1 = len(x) - 1
df2 = len(x) - 1

F = s1_sq / s2_sq
p_value = f.cdf(F, df1, df2)
print(p_value)

0.7584478225464656


#### Since the p-value (0.758) is greater than the significance level of 0.01, we fail to reject the null hypothesis. We do not have sufficient evidence to conclude that the variances of the test scores of the two groups of students are significantly different at the 1% significance level.